[UnifiedPDF][iOS] Update WKFoundTextRange to support found text ranges in PDFs

pxlcoder opened the pull request:
#### dd76ffdd18b08ca780a84e401e7a5eeefb065e5b
<pre>
[UnifiedPDF][iOS] Update WKFoundTextRange to support found text ranges in PDFs
<a href="https://bugs.webkit.org/show_bug.cgi?id=280431">https://bugs.webkit.org/show_bug.cgi?id=280431</a>
<a href="https://rdar.apple.com/136778579">rdar://136778579</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

`WKFoundTextRange` currently only works when searching for DOM text ranges. In
order to support searching PDFs using `UIFindInteraction` / `UITextSearching`,
it should also support retrieving and restoring PDF selections.

To achieve this goal, `WKFoundTextRange` is split into `WKFoundDOMTextRange`
and `WKFoundPDFTextRange`. A PDF selection can be encoded in terms of its
start and end pages, and their respective offsets. `WebFoundTextRange` encodes
the necessary data using a variant.

The PDF ranges still encode frame information, to enable an eventual future
where PDFs in subframes are supported.

Note that `WebFoundTextRangeController` still only knows how to return
and handle DOM text ranges. Subsequent patches will add support for PDF
ranges.

* Source/WebKit/Shared/WebFoundTextRange.cpp:
(WebKit::WebFoundTextRange::hash const):
(WebKit::WebFoundTextRange::operator== const):
* Source/WebKit/Shared/WebFoundTextRange.h:
(WTF::WebFoundTextRangeHash::hash):
* Source/WebKit/Shared/WebFoundTextRange.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView offsetFromPosition:toPosition:]):
(+[WKFoundTextRange foundTextRangeWithWebFoundTextRange:]):
(-[WKFoundTextRange isEmpty]):
(-[WKFoundTextRange webFoundTextRange]):
(-[WKFoundDOMTextRange start]):
(-[WKFoundDOMTextRange end]):
(-[WKFoundDOMTextRange webFoundTextRange]):
(+[WKFoundDOMTextPosition textPositionWithOffset:order:]):
(-[WKFoundPDFTextRange start]):
(-[WKFoundPDFTextRange end]):
(-[WKFoundPDFTextRange webFoundTextRange]):
(+[WKFoundPDFTextPosition textPositionWithPage:offset:]):
(-[WKFoundTextRange start]): Deleted.
(-[WKFoundTextRange end]): Deleted.
(+[WKFoundTextPosition textPositionWithOffset:order:]): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::redraw):
(WebKit::WebFoundTextRangeController::simpleRangeFromFoundTextRange):

Canonical link: <a href="https://commits.webkit.org/284370@main">https://commits.webkit.org/284370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cc2878737f7f60f2dfd83303c71b10123d25b4d2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69125 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21797 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73206 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20283 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56326 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20132 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55014 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13463 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72191 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44291 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59673 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35492 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40959 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18657 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62902 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17448 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74917 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13107 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16687 "Found 1 new test failure: media/video-webkit-playsinline.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62667 "Found 2 new test failures: imported/blink/compositing/animation/hidden-animated-layer-should-not-have-scrollbars.html webanimations/accelerated-animation-tiled-while-running.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13145 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59756 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62566 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10568 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4175 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10565 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44329 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45402 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46598 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Shared/WebFoundTextRange.h:
```diff
@@ -26,17 +26,35 @@
+    };
+
+    struct PDFData {
+        uint64_t startPage { 0 };
```
I could be wrong, but isn't there like a `PageIndex` type for PDFs that would be better suited here?

pxlcoder replied:
There is `PDFDocumentLayout::PageIndex`, but I think it might be layering violation to use here.

I'll defer to Abrar.

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -15048,9 +15093,23 @@ @implementation WKFoundTextRange
-    auto range = adoptNS([[WKFoundTextRange alloc] init]);
-    [range setLocation:webRange.location];
-    [range setLength:webRange.length];
+    RetainPtr range = WTF::switchOn(webRange.data,
```
This looks like one of those thingswhere you could expect to get away without a return type hint, but maybe it was necessary?

pxlcoder replied:
Yeah, the compiler wasn't smart enough to figure out that the two classes were derived from the same class.

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -15062,42 +15121,89 @@ - (void)dealloc
++ (WKFoundDOMTextPosition *)textPositionWithOffset:(NSUInteger)offset order:(NSUInteger)order
 {
-    auto pos = adoptNS([[WKFoundTextPosition alloc] init]);
+    auto pos = adoptNS([[WKFoundDOMTextPosition alloc] init]);
```
Nit: I know this is old code, but `RetainPtr pos = ...`?

pxlcoder replied:
Done.

aprotyas commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -15062,42 +15121,89 @@ - (void)dealloc
+
++ (WKFoundPDFTextPosition *)textPositionWithPage:(NSUInteger)page offset:(NSUInteger)offset
+{
+    auto pos = adoptNS([[WKFoundPDFTextPosition alloc] init]);
```
(Same comment)

pxlcoder replied:
Done.

pxlcoder committed:
[UnifiedPDF][iOS] Update WKFoundTextRange to support found text ranges in PDFs
https://bugs.webkit.org/show_bug.cgi?id=280431
rdar://136778579

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

`WKFoundTextRange` currently only works when searching for DOM text ranges. In
order to support searching PDFs using `UIFindInteraction` / `UITextSearching`,
it should also support retrieving and restoring PDF selections.

To achieve this goal, `WKFoundTextRange` is split into `WKFoundDOMTextRange`
and `WKFoundPDFTextRange`. A PDF selection can be encoded in terms of its
start and end pages, and their respective offsets. `WebFoundTextRange` encodes
the necessary data using a variant.

The PDF ranges still encode frame information, to enable an eventual future
where PDFs in subframes are supported.

Note that `WebFoundTextRangeController` still only knows how to return
and handle DOM text ranges. Subsequent patches will add support for PDF
ranges.

* Source/WebKit/Shared/WebFoundTextRange.cpp:
(WebKit::WebFoundTextRange::hash const):
(WebKit::WebFoundTextRange::operator== const):
* Source/WebKit/Shared/WebFoundTextRange.h:
(WTF::WebFoundTextRangeHash::hash):
* Source/WebKit/Shared/WebFoundTextRange.serialization.in:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView offsetFromPosition:toPosition:]):
(+[WKFoundTextRange foundTextRangeWithWebFoundTextRange:]):
(-[WKFoundTextRange isEmpty]):
(-[WKFoundTextRange webFoundTextRange]):
(-[WKFoundDOMTextRange start]):
(-[WKFoundDOMTextRange end]):
(-[WKFoundDOMTextRange webFoundTextRange]):
(+[WKFoundDOMTextPosition textPositionWithOffset:order:]):
(-[WKFoundPDFTextRange start]):
(-[WKFoundPDFTextRange end]):
(-[WKFoundPDFTextRange webFoundTextRange]):
(+[WKFoundPDFTextPosition textPositionWithPage:offset:]):
(-[WKFoundTextRange start]): Deleted.
(-[WKFoundTextRange end]): Deleted.
(+[WKFoundTextPosition textPositionWithOffset:order:]): Deleted.
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::redraw):
(WebKit::WebFoundTextRangeController::simpleRangeFromFoundTextRange):

Canonical link: https://commits.webkit.org/284370@main
